File: packages/multifatorial/android/src/main/java/br/com/voxdatati/voxauthflutter/VoxauthflutterPlugin.java
package br.com.voxdatati.voxauthflutter;

import android.app.Activity;
import android.os.AsyncTask;

import androidx.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import br.com.voxdatati.voxauth.OTP.Util.CryptoAlgorithms;
import br.com.voxdatati.voxauth.SDK.AuthenticationMethod;
import br.com.voxdatati.voxauth.SDK.Callback.GetDeviceInfoCallBack;
import br.com.voxdatati.voxauth.SDK.CodReturn;
import br.com.voxdatati.voxauth.SDK.DeviceInfo;
import br.com.voxdatati.voxauth.SDK.Exception.GenericException;
import br.com.voxdatati.voxauth.SDK.Return;
import br.com.voxdatati.voxauth.SDK.SDKClient;
import br.com.voxdatati.voxauth.SDK.Transaction;


/** VoxauthflutterPlugin */
public class VoxauthflutterPlugin  implements FlutterPlugin, MethodCallHandler, ActivityAware {

    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    public Activity activity;
    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "voxauthflutter");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        activity = binding.getActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {

    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

    }

    @Override
    public void onDetachedFromActivity() {

    }


    public class  voxAuthFlutter {

        private SDKClient client = null;
        private Transaction transaction = new Transaction();
        private DeviceInfo info = null;
        private String WebService = "";
        private String Channel = "";
        private String Password = "";
        private String ClientId = "";
        private String DeviceName = "";

        public void   voxAuthFlutter(String webServiceUrl, String channelName, String password, String clientId, String deviceName, Result result) {

            try {
                if (client == null || WebService != webServiceUrl || ClientId != clientId || Password != password || DeviceName != deviceName) {
                    WebService = webServiceUrl;
                    ClientId = clientId;
                    Password = password;
                    DeviceName = deviceName;
                    info = null;
                    ExecutorService executor = Executors.newSingleThreadExecutor();
                    executor.execute(() -> {
                        client = new SDKClient(webServiceUrl, channelName, password, clientId, deviceName);
                        client.SyncDate();
                    });
                    executor.shutdown();
                }
               result.success("ok");

            } catch (final Exception exception) {
                    result.error("400",exception.getMessage(), exception );
            }
        }

        public void  GetDeviceInfo(Result result) {
            try {
                    if (info == null) {
                        ExecutorService executor = Executors.newSingleThreadExecutor();
                        executor.execute(() -> client.GetDeviceInfo(activity, new GetDeviceInfoCallBack() {
                            @Override
                            public void OnSuccess(DeviceInfo deviceInfo) {
                                List IMEI = Arrays.asList();
                                for (Object i : deviceInfo.IMEI) {
                                    IMEI.add(i);
                                }
                                JSONObject res = new JSONObject();
                                try {
                                    res.put("Board", deviceInfo.Board);
                                    res.put("Device", deviceInfo.Device);
                                    res.put("Brand", deviceInfo.Brand);
                                    res.put("MAC", deviceInfo.MAC);
                                    res.put("Latitude", deviceInfo.Latitude);
                                    res.put("Longitude", deviceInfo.Longitude);
                                    res.put("Model", deviceInfo.Model);
                                    res.put("NicknameDevice", deviceInfo.NicknameDevice);
                                    res.put("RegistrationAppID", deviceInfo.RegistrationAppID);
                                    res.put("SerialNumber", deviceInfo.SerialNumber);
                                    res.put("Software", deviceInfo.Software);
                                    res.put("SoftwareEmulator", deviceInfo.SoftwareEmulator);
                                    res.put("SoftwareVersion", deviceInfo.SoftwareVersion);
                                    res.put("IMEI", IMEI);
                                    res.put("SoftwareRooted", deviceInfo.SoftwareRooted);
                                    info = deviceInfo;
                                    result.success(res.toString());
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                }
                            }

                            @Override
                            public void OnFail(GenericException exception) {
                                result.error("400", "notOk", exception);
                            }
                        }));
                        executor.shutdown();
                    }
                else {
                    Exception exception = new Exception("Device info has already been loaded");
                    result.error("400","Device info has already been loaded", exception);
                }
            } catch (final Exception exception) {
                result.error("400",exception.getMessage(), exception);
            }
        }

        public void GetDeviceID (Result result) {
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        String DeviceID = client.GetDeviceID(activity);
                        if (DeviceID == "ok") {
                        }
                        result.success(DeviceID);
                    } catch (final Exception exception) {
                        result.error("400", exception.getMessage(), exception);
                    }
                }
            });
            executor.shutdown();
        }

        public void SyncAuthenticationMethodsConfigs(Result result) {
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    if
                    (info != null) {
                        List<AuthenticationMethod> methods = client.SyncAuthenticationMethodsConfigs(info);
                        List<JSONObject> jsonArray = new ArrayList<JSONObject>();
                        for (AuthenticationMethod authMethod : methods) {
                            String criptoString = "";
                            JSONObject jsonObject = new JSONObject();
                            if (authMethod.Cryptography != null) {
                                criptoString = authMethod.Cryptography.toString();
                            } else if (authMethod.Cryptography == null) {
                                criptoString = "";
                            }
                            jsonObject.put("NameAuthenticationMethodType", authMethod.NameAuthenticationMethodType);
                            jsonObject.put("CanCreateNew", authMethod.CanCreateNew);
                            jsonObject.put("IDAuthenticationMethod", authMethod.IDAuthenticationMethod);
                            jsonObject.put("DisplayNumber", authMethod.DisplayNumber);
                            jsonObject.put("Status", authMethod.Status);
                            jsonObject.put("Cryptography", criptoString);
                            jsonObject.put("Key", authMethod.Key);
                            jsonObject.put("PassSize", authMethod.PassSize);
                            jsonObject.put("PassGenerationTime", authMethod.PassGenerationTime);
                            jsonArray.add(jsonObject);
                        }
                        result.success(jsonArray.toString());
                    } else {
                        Exception exception = new Exception("No device info");
                        result.error("400", exception.getMessage(), exception);
                    }
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void InsertMobileToken (Result result) {
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    if (info != null) {
                        AuthenticationMethod method = client.InsertMobileToken(info);
                        JSONObject jsonObject = new JSONObject();
                        jsonObject.put("NameAuthenticationMethodType", method.NameAuthenticationMethodType);
                        jsonObject.put("CanCreateNew", method.CanCreateNew);
                        jsonObject.put("IDAuthenticationMethod", method.IDAuthenticationMethod);
                        result.success(jsonObject.toString());
                    } else {
                        Exception exception = new Exception("No device info");
                        result.error("400", exception.getMessage(), exception);
                    }
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void GetPassword (String Cryptography, String Key, int PassSize, int PassGenerationTime,Result result) {
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    CryptoAlgorithms CryptographyEnum = CryptoAlgorithms.valueOf(Cryptography);
                    String token = client.GetPassword(CryptographyEnum, Key, PassSize, PassGenerationTime);
                    result.success(token);
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void GetAuthenticationMethodsMobile(String transactionCode,Double transactionValue, Result result) {
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    Transaction transaction = new Transaction();
                    transaction.setCode(transactionCode);
                    transaction.setDate(Calendar.getInstance().getTime());
                    transaction.setValue(transactionValue);
                    List<AuthenticationMethod> methods = client.GetAuthenticationMethodsMobile(info, transaction);
                    List<JSONObject> jsonArray = new ArrayList<JSONObject>();
                    for (AuthenticationMethod authMethod : methods) {
                        JSONObject jsonObject = new JSONObject();
                        jsonObject.put("NameAuthenticationMethodType", authMethod.NameAuthenticationMethodType);
                        jsonObject.put("CanCreateNew", authMethod.CanCreateNew);
                        jsonObject.put("IDAuthenticationMethod", authMethod.IDAuthenticationMethod);
                        jsonObject.put("DisplayNumber", authMethod.DisplayNumber);
                        jsonObject.put("Status", authMethod.Status);
                        jsonArray.add(jsonObject);
                    }
                    result.success(jsonArray.toString());
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void MTokenAuthentication(String transactionCode,Double transactionValue,Double IDAuthenticationMethod, String token ,Result result){
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    Transaction transaction = new Transaction();
                    transaction.setCode(transactionCode);
                    transaction.setDate(Calendar.getInstance().getTime());
                    transaction.setValue(transactionValue);
                    Return TransactionReturn = client.MTokenAuthentication(transaction, IDAuthenticationMethod.longValue(), token);
                    if (TransactionReturn.CodReturn == CodReturn.SUCCESS) {
                        result.success(Integer.toString(TransactionReturn.CodReturn));
                    } else {
                        String errorMessage = TransactionReturn.MessageReturn;
                        int codRet = TransactionReturn.CodReturn; // Verificar a classe CodReturn
                        Exception exception = new Exception(errorMessage + " - " + codRet);
                        result.error("400", exception.getMessage(), exception);
                    }
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void SMSAuthentication(String transactionCode, Double transactionValue, Double IDAuthenticationMethod, String telephone, Result result){
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    Transaction transaction = new Transaction();
                    transaction.setCode(transactionCode);
                    transaction.setDate(Calendar.getInstance().getTime());
                    transaction.setValue(transactionValue);
                    Return SMSAuthReturn = client.SMSAuthentication(transaction, IDAuthenticationMethod.longValue(), telephone);
                    JSONObject jsonObject = new JSONObject();
                    jsonObject.put("CodReturn", SMSAuthReturn.CodReturn);
                    jsonObject.put("AuthenticationSignature", SMSAuthReturn.AuthenticationSignature);
                    jsonObject.put("MessageReturn", SMSAuthReturn.MessageReturn);
                    result.success(jsonObject.toString());
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void ValidateAuthentication(String token, String authenticationSignature, Result result){
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    Return validateAuthReturn = client.ValidateAuthentication(authenticationSignature, token);
                    result.success(Integer.toString(validateAuthReturn.CodReturn));
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void ReverseSMSAuthentication(String transactionCode, Double transactionValue, Double IDAuthenticationMethod, String telephone, Result result){
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    Transaction transaction = new Transaction();
                    transaction.setCode(transactionCode);
                    transaction.setDate(Calendar.getInstance().getTime());
                    transaction.setValue(transactionValue);
                    Return SMSAuthReturn = client.ReverseSMSAuthentication(transaction, IDAuthenticationMethod.longValue(), telephone);

                    JSONObject jsonObject = new JSONObject();
                    jsonObject.put("CodReturn", SMSAuthReturn.CodReturn);
                    jsonObject.put("AuthenticationSignature", SMSAuthReturn.AuthenticationSignature);
                    jsonObject.put("MessageReturn", SMSAuthReturn.MessageReturn);
                    result.success(jsonObject.toString());

                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }

        public void ResponseAuthentication(String authenticationSignature, Result result){
            ExecutorService executor = Executors.newSingleThreadExecutor();
            executor.execute(() -> {
                try {
                    Return responseAuthReturn = client.ResponseAuthentication(authenticationSignature);
                    result.success(Integer.toString(responseAuthReturn.CodReturn));
                } catch (final Exception exception) {
                    result.error("400", exception.getMessage(), exception);
                }
            });
            executor.shutdown();
        }
    }

    private static voxAuthFlutter SDK;

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("getPlatformVersion")) {
            result.success("Android " + android.os.Build.VERSION.RELEASE);
        }
        else if (call.method.equals("SDKClient")) {
            String webServiceUrl = call.argument("webServiceUrl");
            String channelName = call.argument("channelName");
            String password = call.argument("password");
            String clientId  = call.argument("clientId");
            String deviceName = call.argument("deviceName");
            SDK = new voxAuthFlutter();
            SDK.voxAuthFlutter(webServiceUrl,channelName, password, clientId, deviceName, result);
        }
        else if (call.method.equals("GetDeviceInfo")){
            SDK.GetDeviceInfo(result);
        }
        else if (call.method.equals("GetAuthenticationMethodsMobile")){
            String transactionCode = call.argument("transactionCode");
            Double transactionValue = call.argument("transactionValue");
            SDK.GetAuthenticationMethodsMobile(transactionCode,transactionValue,result);
        }
        else if (call.method.equals("GetDeviceID")){
            SDK = new voxAuthFlutter();
            SDK.GetDeviceID(result);
        }
        else if (call.method.equals("SyncAuthenticationMethodsConfigs")){
            SDK.SyncAuthenticationMethodsConfigs(result);
        }
        else if (call.method.equals("InsertMobileToken")) {
            SDK.InsertMobileToken(result);
        }
        else if (call.method.equals("GetPassword")){
            String Cryptography = call.argument("Cryptography");
            String Key = call.argument("Key");
            Integer PassSize = call.argument("PassSize");
            Integer PassGenerationTime  = call.argument("PassGenerationTime");
            SDK.GetPassword(Cryptography, Key, PassSize, PassGenerationTime, result);
        }
        else if (call.method.equals("MTokenAuthentication")) {
            String transactionCode = call.argument("transactionCode");
            Double transactionValue = call.argument("transactionValue");
            Double IDAuthenticationMethod = call.argument("IDAuthenticationMethod");
            String token  = call.argument("token");
            SDK.MTokenAuthentication(transactionCode, transactionValue, IDAuthenticationMethod, token, result);
        }
        else if (call.method.equals("SMSAuthentication")) {
            String transactionCode = call.argument("transactionCode");
            Double transactionValue = call.argument("transactionValue");
            Double IDAuthenticationMethod = call.argument("IDAuthenticationMethod");
            String phone  = call.argument("phone");
            SDK.SMSAuthentication(transactionCode, transactionValue, IDAuthenticationMethod, phone, result);
        }
        else if (call.method.equals("ValidateAuthentication")) {
            String token = call.argument("token");
            String authenticationSignature  = call.argument("authenticationSignature");
            SDK.ValidateAuthentication(token, authenticationSignature, result);
        }
        else if (call.method.equals("ReverseSMSAuthentication")) {
            String transactionCode = call.argument("transactionCode");
            Double transactionValue = call.argument("transactionValue");
            Double IDAuthenticationMethod = call.argument("IDAuthenticationMethod");
            String phone  = call.argument("phone");
            SDK.ReverseSMSAuthentication(transactionCode, transactionValue, IDAuthenticationMethod, phone, result);
        }
        else if (call.method.equals("ResponseAuthentication")) {
            String authenticationSignature  = call.argument("authenticationSignature");
            SDK.ResponseAuthentication(authenticationSignature, result);
        }
        else {
            result.notImplemented();
        }
    }
    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

}

File: packages/multifatorial/ios/Classes/VoxauthflutterPlugin.swift
import Flutter
import UIKit
import VoxAuth

public class VoxauthflutterPlugin: NSObject, FlutterPlugin {
    var client:VoxAuth.VOXAUTHSDKClient!;
    var transation:Transaction! = Transaction();
    var deviceInfo:DeviceInfo!;
    var WebService:String = "";
    var Channel:String = "";
    var Password:String = "";
    var ClientId:String = "";
    var DeviceName:String = "";

    func SDKClient(webServiceUrl: String, channelName: String ,password: String, clientId: String, deviceName: String, result: FlutterResult) -> Void  {
            do {
                if (client == nil || WebService != webServiceUrl || ClientId != clientId || Password != password || DeviceName != deviceName) {
                    WebService = webServiceUrl;
                    ClientId = clientId;
                    Password = password;
                    DeviceName = deviceName;
                    Channel = channelName;
                    // info = null;
                    client = VoxAuth.VOXAUTHSDKClient(webServiceUrl: webServiceUrl, channelName: channelName, password: password, clientId: clientId, nicknameDevice: deviceName);
                    //try client.syncDate();
                }else{
                    throw "SDKClient Error"
                }
                
                result("ok");
                
            } catch let error {
                result(
                        FlutterError( code: "invalidArgs",
                        message: "Erro SDKClient",
                        details: error.localizedDescription )
                )
            }
        }
    
    func GetDeviceInfo(result: FlutterResult){
           do{
               deviceInfo = try client.getDeviceInfo();
               if(deviceInfo != nil){
                   deviceInfo.set(nickname: DeviceName);
                   result("ok");
               }else{
                   throw "GetDeviceInfo Error"
               }
           } catch let error {
               result(
                       FlutterError( code: "invalidArgs",
                       message: "Erro GetDeviceInfo",
                       details: error.localizedDescription )
               )
           }
       }
    
    func GetDeviceID(result: FlutterResult){
                do{
                    let deviceID: String = try VoxAuth.VOXAUTHSDKClient.getDeviceID();
                        result(deviceID);
                } catch let error {
                    result(
                            FlutterError( code: "invalidArgs",
                            message: "Erro GetDeviceID",
                            details: error.localizedDescription )
                    )
                 }
            }
    
    func SyncAuthenticationMethodsConfigs(result: FlutterResult){
            var dictionaryMethods:Array<Any> = []
                do{
                    let methods: [AuthenticationMethod] = try client.syncAuthenticationMethodsConfigs(info: deviceInfo)
                    
                    struct STAuthenticationMethod: Codable {
                        var IDAuthenticationMethod : Int64 = 0
                        var IDAuthenticationMethodType : Int = 0
                        var NameAuthenticationMethodType : String = ""
                        var Validity : String = ""
                        var PassSize : Int = 0
                        var PassGenerationTime : Int = 0
                        var Cryptography : String = ""
                        var DisplayNumber : String = ""
                        var Key : String = ""
                        var Status : String = ""
                        var ProjectID : String = ""
                        var Percentual : UInt8 = 0
                        var CanCreateNew : Bool = false
                        var IDMobileDevice : String = ""
                    }
                    
                    for m in methods {
                        
                        let instance = STAuthenticationMethod(IDAuthenticationMethod: m.IDAuthenticationMethod, IDAuthenticationMethodType: m.IDAuthenticationMethodType, NameAuthenticationMethodType: m.NameAuthenticationMethodType, Validity: m.Validity, PassSize: Int(m.PassSize) ?? 0, PassGenerationTime: Int(m.PassGenerationTime) ?? 0, Cryptography: m.Cryptography, DisplayNumber: m.DisplayNumber, Key: m.Key, Status: m.Status, ProjectID: m.ProjectID, Percentual: m.Percentual, CanCreateNew: m.CanCreateNew, IDMobileDevice: m.IDMobileDevice)
                        
                        
                        let jsonEncoder = JSONEncoder()
                        jsonEncoder.outputFormatting = .prettyPrinted
                        let jsonData:Data = try jsonEncoder.encode(instance)
                        let json = try JSONSerialization.jsonObject(with: jsonData, options: [])
    
                        
                        dictionaryMethods.append(json)
                        
                    }
                    	let jsonString = json(from: dictionaryMethods)!
                    result(jsonString);
                    
                } catch let error {
                    result(
                            FlutterError( code: "invalidArgs",
                            message: "Erro SyncAuthenticationMethodsConfigs",
                            details: error.localizedDescription )
                    )
                }
            }
    func json(from object:Any) -> String? {
                guard let data = try? JSONSerialization.data(withJSONObject: object, options: []) else {
                    return nil
                }
                return String(data: data, encoding: String.Encoding.utf8)
            }


    func InsertMobileToken(result: FlutterResult){
            do{
                let methods: AuthenticationMethod = try client.insertMobileToken(info: deviceInfo)

                
                struct STAuthenticationMethod: Codable {
                       var NameAuthenticationMethodType : String = ""
                       var CanCreateNew : Bool = false
                       var IDAuthenticationMethod : Int64 = 0
                }
                let instance = STAuthenticationMethod(NameAuthenticationMethodType: methods.NameAuthenticationMethodType,
                                 CanCreateNew: methods.CanCreateNew,
                                 IDAuthenticationMethod: methods.IDAuthenticationMethod)

                             let jsonEncoder = JSONEncoder()
                             jsonEncoder.outputFormatting = .prettyPrinted
                             let jsonData:Data = try jsonEncoder.encode(instance)
                             let jsonString =  String(data: jsonData, encoding: String.Encoding.utf8);
                            result(jsonString);

            } catch let error {
                result(
                        FlutterError( code: "invalidArgs",
                        message: "Erro InsertMobileToken",
                        details: error.localizedDescription )
                );
            }
        }
    
    func GetPassword(Cryptography:String ,Key:String ,PassSize:String ,PassGenerationTime:String , result: FlutterResult){
                do{
                    let method:AuthenticationMethod! = AuthenticationMethod()
                    method.Cryptography = Cryptography;
                    method.Key = Key;
                    method.PassSize = PassSize;
                    method.PassGenerationTime = PassGenerationTime;

                    let token:String = try client.getMobileToken(algorithm: method.getOTPAlgorithm(), secret: method.Key, size: method.PassSize, interval: method.PassGenerationTime)
                    result(token)
                } catch let error {
                    result(
                            FlutterError( code: "invalidArgs",
                            message: "Erro GetPassword",
                            details: error.localizedDescription )
                    );
                }
            }
    
    func GetAuthenticationMethodsMobile(transactionCode:String, transactionValue:Float , result: FlutterResult){
            var dictionaryMethods:Array<Any> = []
            do{
                let transation:Transaction = Transaction();
                
                transation.Code = transactionCode;
                transation.Value = Double(transactionValue);
                
                let methods: [AuthenticationMethod] = try client.getAuthenticationMethodsMobile(info: deviceInfo,transaction: transation)

                struct STAuthenticationMethod: Codable {
                    var NameAuthenticationMethodType : String = ""
                    var CanCreateNew : Bool = false
                    var IDAuthenticationMethod : Int64 = 0
                    var DisplayNumber : String = ""
                    var Status : String = ""
                }

                for m in methods {
                    let instance = STAuthenticationMethod(NameAuthenticationMethodType: m.NameAuthenticationMethodType,
                                                            CanCreateNew: m.CanCreateNew,
                                                            IDAuthenticationMethod: m.IDAuthenticationMethod,
                                                            DisplayNumber: m.DisplayNumber,
                                                            Status: m.Status)

                    let jsonEncoder = JSONEncoder()
                    jsonEncoder.outputFormatting = .prettyPrinted
                    let jsonData:Data = try jsonEncoder.encode(instance)
                    let jsonString = try JSONSerialization.jsonObject(with: jsonData, options: [])

                    dictionaryMethods.append(jsonString)

                    }
               let jsonResult = json(from: dictionaryMethods)!
                result(jsonResult);

                } catch let error {
                    result(
                            FlutterError( code: "invalidArgs",
                            message: "Erro GetAuthenticationMethodsMobile",
                            details: error.localizedDescription )
                    );
                }
        }

    
    func MTokenAuthentication(transactionCode:String, transactionValue:Float, IDAuthenticationMethod:Float, token:String, result: FlutterResult){
                do{
                    let transation:Transaction = Transaction();
                    
                    transation.Code = transactionCode;
                    
                    transation.Value = Double(transactionValue);
                    
                    let method:AuthenticationMethod! = AuthenticationMethod();
                    
                    let idInt = Int(IDAuthenticationMethod);
                    
                    let idInt64 = Int64(exactly: idInt);
                    
                    method.IDAuthenticationMethod   = idInt64!;

                    let transactionReturn = try client.MTokenAuthentication(transaction: transation, idAuthenticationMethod: method.IDAuthenticationMethod, mtoken: token)
                    
                    result(transactionReturn.CodReturn.description)

                } catch let error {
                    result(
                            FlutterError( code: "invalidArgs",
                            message: "Erro MTokenAuthentication",
                            details: error.localizedDescription )
                    );
                }
            }
    
    func SMSAuthentication(transactionCode:String, transactionValue:Float, IDAuthenticationMethod:Float, telephone:String,result: FlutterResult){
            
            struct STReturnSMS: Codable {
                var CodReturn : Int;
                var AuthenticationSignature : String;
            }
            
            do{
                let transation:Transaction = Transaction();
                
                transation.Code = transactionCode;
                
                transation.Value = Double(transactionValue);
                
                let method:AuthenticationMethod! = AuthenticationMethod();
                
                let idInt = Int(IDAuthenticationMethod);
                
                let idInt64 = Int64(exactly: idInt);
                
                method.IDAuthenticationMethod   = idInt64!;

                let transactionReturn = try client.SMSAuthentication(transaction: transation, idAuthenticationMethod: method.IDAuthenticationMethod, telephone: telephone)
                
                let returnSMS = STReturnSMS(CodReturn: transactionReturn.CodReturn ,AuthenticationSignature: transactionReturn.AuthenticationSignature )

                let jsonEncoder = JSONEncoder()
                
                jsonEncoder.outputFormatting = .prettyPrinted
                
                let jsonData: Data = try jsonEncoder.encode(returnSMS)
                
                let jsonString =  String(data: jsonData, encoding: String.Encoding.utf8);
                 result(jsonString);

            } catch let error {
                result(
                        FlutterError( code: "invalidArgs",
                        message: "Erro SMSAuthentication",
                        details: error.localizedDescription )
                );
            }
        }

    func ValidateAuthentication(token: String, authenticationSignature: String, result: FlutterResult){
            do{
                let validationReturn = try client.validateAuthentication(authSignature: authenticationSignature, passwordReceived: token )

                result(validationReturn.CodReturn.description)

            } catch let error {
                result(
                        FlutterError( code: "invalidArgs",
                        message: "Erro ValidateAuthentication",
                        details: error.localizedDescription )
                );
            }
        }
    
    func ReverseSMSAuthentication(transactionCode:String, transactionValue:Float, IDAuthenticationMethod:Float, telephone:String, result: FlutterResult){
           
           struct STReturnReverseSMS: Codable {
               var CodReturn : Int;
               var AuthenticationSignature : String;
           }
           
           do{
               let transation:Transaction = Transaction();
               
               transation.Code = transactionCode;
               
               transation.Value = Double(transactionValue);
               
               let method:AuthenticationMethod! = AuthenticationMethod();
               
               let idInt = Int(IDAuthenticationMethod);
               
               let idInt64 = Int64(exactly: idInt);
               
               method.IDAuthenticationMethod   = idInt64!;

               let transactionReturn = try client.ReverseSMSAuthentication(transaction: transation, idAuthenticationMethod: method.IDAuthenticationMethod, telephone: telephone)
               
               let returnSMS = STReturnReverseSMS(CodReturn: transactionReturn.CodReturn ,AuthenticationSignature: transactionReturn.AuthenticationSignature )

               let jsonEncoder = JSONEncoder()
               
               jsonEncoder.outputFormatting = .prettyPrinted
               
               let jsonData: Data = try jsonEncoder.encode(returnSMS)
               
               let jsonString =  String(data: jsonData, encoding: String.Encoding.utf8);
                result(jsonString);

           } catch let error {
               result(
                       FlutterError( code: "invalidArgs",
                       message: "Erro ReverseSMSAuthentication",
                       details: error.localizedDescription )
               );
           }
       }
    
    func ResponseAuthentication(authenticationSignature: String, result: FlutterResult){
            do{
                let authenticationReturn = try client.ResponseAuthentication(authSignature: authenticationSignature)

                result(authenticationReturn.CodReturn.description)

            } catch let error {
                result(
                        FlutterError( code: "invalidArgs",
                        message: "Erro ResponseAuthentication",
                        details: error.localizedDescription )
                );
            }
        }







          

    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "voxauthflutter", binaryMessenger: registrar.messenger())
    let instance = VoxauthflutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method { 
        case "SDKClient":
        if let args = call.arguments as? Dictionary<String, Any>,
           let webServiceUrl = args["webServiceUrl"] as? String,
           let password = args["password"] as? String,
           let clientId = args["clientId"] as? String,
           let deviceName = args["deviceName"] as? String,
           let channelName = args["channelName"] as? String {
            SDKClient(webServiceUrl: webServiceUrl,channelName: channelName, password: password,clientId: clientId, deviceName: deviceName,result: result);
         } else {
           result(FlutterError.init(code: "bad args", message: nil, details: nil))
         }
    case "GetDeviceInfo":
        GetDeviceInfo(result: result);
    case "GetDeviceID":
        GetDeviceID(result: result);
    case "SyncAuthenticationMethodsConfigs":
        SyncAuthenticationMethodsConfigs(result: result);
    case "InsertMobileToken":
        InsertMobileToken(result: result);
    case "GetPassword":
            if let args = call.arguments as? Dictionary<String, Any>,
               let Cryptography = args["Cryptography"] as? String,
               let PassGenerationTime = args["PassGenerationTime"] as? Int32,
               let PassSize = args["PassSize"] as? Int32,
               let Key = args["Key"] as? String
                    {
                    let auxPass:String = String(PassSize);
                    let auxTime:String = String(PassGenerationTime)
                    GetPassword(Cryptography: Cryptography, Key: Key, PassSize: auxPass, PassGenerationTime: auxTime, result: result);
                }
        
    case "GetAuthenticationMethodsMobile":
            if let args = call.arguments as? Dictionary<String, Any>,
               let transactionCode = args["transactionCode"] as? String,
               let transactionValue = args["transactionValue"] as? Float
                    {
                GetAuthenticationMethodsMobile(transactionCode: transactionCode, transactionValue: transactionValue, result: result);
                }
    case "MTokenAuthentication":
            if let args = call.arguments as? Dictionary<String, Any>,
               let transactionCode = args["transactionCode"] as? String,
               let transactionValue = args["transactionValue"] as? Float,
               let IDAuthenticationMethod = args["IDAuthenticationMethod"] as? Float,
               let token = args["token"] as? String
                    {
                MTokenAuthentication(transactionCode: transactionCode, transactionValue: transactionValue, IDAuthenticationMethod: IDAuthenticationMethod, token: token, result: result);
                }
    case "SMSAuthentication":
            if let args = call.arguments as? Dictionary<String, Any>,
               let transactionCode = args["transactionCode"] as? String,
               let transactionValue = args["transactionValue"] as? Float,
               let IDAuthenticationMethod = args["IDAuthenticationMethod"] as? Float,
               let telephone = args["phone"] as? String
                    {
                SMSAuthentication(transactionCode: transactionCode, transactionValue: transactionValue, IDAuthenticationMethod: IDAuthenticationMethod, telephone: telephone, result: result);
                }
    case "ValidateAuthentication":
            if let args = call.arguments as? Dictionary<String, Any>,
               let token = args["token"] as? String,
               let authenticationSignature = args["authenticationSignature"] as? String
                    {
                ValidateAuthentication(token: token, authenticationSignature: authenticationSignature, result: result);
                }
        
    case "ReverseSMSAuthentication":
            if let args = call.arguments as? Dictionary<String, Any>,
               let transactionCode = args["transactionCode"] as? String,
               let transactionValue = args["transactionValue"] as? Float,
               let IDAuthenticationMethod = args["IDAuthenticationMethod"] as? Float,
               let telephone = args["phone"] as? String
                    {
                ReverseSMSAuthentication(transactionCode: transactionCode, transactionValue: transactionValue, IDAuthenticationMethod: IDAuthenticationMethod, telephone: telephone, result: result);
                }
    case "ResponseAuthentication":
            if let args = call.arguments as? Dictionary<String, Any>,
               let authenticationSignature = args["authenticationSignature"] as? String
                    {
                ResponseAuthentication(authenticationSignature: authenticationSignature, result: result);
                }

        
    default:
            result("iOS " + UIDevice.current.systemVersion) 
    }
  }
}
	
